### ISSUE TITLE
Rust analyzer reports errors not reported by rustc

### ISSUE THREAD

**Issue by DanielT** [rust-lang/rust-analyzer#22454]
Rust analyzer version: 0.3.2913 (Linux)
rustc version: 1.95.0

I have a program that causes rust-analyzer to report errors where rustc does not.

I've reduced my program to a short repro example:
```Rust
// lib.rs
pub fn testcase(xml_str: &str) {
    let mut reader = quick_xml::Reader::from_str(xml_str);
    match reader.read_event().unwrap() {
        quick_xml::events::Event::GeneralRef(bytes_ref) => {
            match bytes_ref.as_ref() {
                b"amp" => {}
                b"lt" => {}
                b"apos" => {}
                _ => {}
            };
        }
        _ => {}
    }
}
```
```Toml
# Cargo.toml
[package]
name = "ra-bug-repro"
edition = "2024"

[dependencies]
quick-xml = "0.40.1"
zip = "8.6.0"
```
Rust analyzer reports:
```
Line 7: expected &[u8; 3], found &'static [u8; 2]
Line 8: expected &[u8; 3], found &'static [u8; 4]
```

I made a particularly interesting discovery while reducing my code to the testcase: If the dependency on `zip` is removed in `Cargo.toml` then rust-analyzer no longer reports any errors. Just including this dependency, even unused, changes the behavior.

**Comment by ChayimFriedman2** [comment]
Adding `extern crate zip;` at the top makes rustc also report an error. So the problem is some `AsRef` impl in `zip`, and that rustc doesn't link the crate if it's unused and there is no `extern crate` for it while we do.

This is a bit of an edge case and while I'm not sure yet why we link the crate, we might decide to not fix this if it turns out too hard.

**Comment by DanielT** [comment]
Thanks for the quick response.

In that case I'll constrain the trait inference by changing my code:
```Rust
...
let text: &[u8] = bytes_ref.as_ref()
match text {
...
```

As far as I'm concerned the matter is solved. I don't know if you want to keep this issue open to track down the crate linking edge case so I haven't closed it.

**Comment by Hmikihiro** [comment]
This error can also be reproduced without zip or quick-xml 
```
.
├── Cargo.toml
├── src
│   └── lib.rs
└── sub
    ├── Cargo.toml
    └── src
        └── lib.rs
```

```rs
// src/lib.rs
+extern crate sub;

pub fn testcase() {
    let byte_ref: &[u8] = b"abcd";
    match byte_ref.as_ref() {
        b"amp" => {}
        b"lt" => {}
        b"apos" => {}
        _ => {}
    };
}
```
```toml
# Cargo.toml
[package]
name = "ra-bug-repro"
edition = "2024"

[dependencies]
sub = { path = "./sub" }
```

```rs
// sub/src/lib.rs
pub struct Path;

impl AsRef<Path> for [u8] {
    fn as_ref(&self) -> &Path {
        &Path
    }
}
```

```toml
# sub/Cargo.toml
[package]
name = "sub"
edition = "2024"

[dependencies]
```

**Comment by CinisBorn** [comment]
I do not know If it is related, but I forked tokio repository, then an error was triggered to `taskdump`, an unstable feature of tokio, but I do not activated anything. 

Interesting fact: `cargo test` and `cargo check workspace` works fine.

**Comment by ChayimFriedman2** [comment]
@CinisBorn Unlikely to be related.

**Comment by ChayimFriedman2** [comment]
So looking at how rustc implements dynamic registration of crates, we indeed cannot implement that, so this is a won't-fix. Closing.
